docs/ECN-test-plan.md

```diff
@@ -0,0 +1,151 @@
+        inter-frame transmission interval should be lesser than
+        per-frame pause duration.
+
+#### Test Steps
+
+1. Start PFC PAUSE storm to fully block the lossless priority at the
+    DUT.
+2. From the Keysight Tx port, send 2N data packets to the receiver. The packets should be mapped to priority Pi on the DUT.
+3. After all the test data packets are sent, start data capture on the Keysight Rx port. Then, stop PFC PAUSE.
+4. Stop capture after all packets are received.
+5. Verify the following:
+   * Keysight Rx port must receive 2N test data packets.
+   * The first test data packet received by Keysight Rx port should be ECN marked.
+   * The last test data packet received by Keysight Rx port should not be ECN marked.
+5. Repeat the test with a different lossless priority (!=Pi).
+
+
+### Test Case #2 - ECN marking accuracy
+
+#### Test Objective
+
+This test aims to verify the DUT’s dequeue based ECN marking behavior (Egress marking) for data packets configured with lossless priority, when minimum threshold < maximum threshold.
+
+#### Test Configuration
+
+- On SONiC DUT configure the following:
+  1. A single lossless priority value Pi (0 <= i <= 7).
+  2. Configure the minimum ECN marking threshold, the maximum ECN marking threshold, and the maximum marking probability of the Priority Pi to Kmin KB, Kmax KB, and Pmax % respectively. Let the Kmin, Kmax, and Pmax be 500 KB, 2000 KB, and 5% for first iteration.
+  3. DUT should have enough buffer to hold (Kmax+10) packets each of 1KB size.
+
+- Configure following traffic items on the Keysight device:
+  1. Test data traffic: A traffic item from the Keysight Tx port to
+        the Keysight Rx port with the lossless priority (DSCP value == Pi).
+        Should have fixed number of packets, Kmax+10 (2010), each having 1KB frame size.
+  2. PFC PAUSE storm: Persistent PFC pause frames from the Keysight
+        Rx port to the Keysight Tx port. The priorities of PFC pause
+        frames should be the same as that of 'Test data traffic'. And the
+        inter-frame transmission interval should be lesser than the 
+        per-frame pause duration.
+
+#### Test Steps
+
+1. Start PFC PAUSE storm to fully block the lossless priority at the
+    DUT.
+2. Start the Test data traffic, it will send total 2010 packets as per the configuration.
+3. After all the test data packets are sent, start data capture on the Keysight Rx port. Then, stop PFC PAUSE storm from Keysight Rx port.
+4. Stop capture after all packets are received.
+5. Verify the following:
+   * All 2010 packets must be received at the Rx port.
+   * As per RED ECN egress marking algorithm, the queue length associated with the data packet i (i = 1, 2, … Kmax + 10) is (Kmax + 10 – i) KB.
+   * If instantaneous queue length is q, then, for the first 10 packets, q >= 2000. Hence, all the first 10 packets should be ECN marked.
+   * For the next 1500 packets, 500 <= q < 2000, so, 5% of these packets should be ECN marked.
```

Comment:
For the next Kmax - Kmin packets, Kmin <= q < Kmax. Thus their marking probabilities should be smaller than Pmax.

Reply:
According to the RED ECN algorithm it should linearly vary from Pmax to 0 as queue_length reduces from Kmax to Kmin. I updated accordingly.